app/models/user.rb

```diff
@@ -252,12 +253,43 @@ def last_activities(last_activity_id = nil, per_page = 10)
       .uniq
   end
 
+  def self.find_by_valid_wopi_token(token)
+    Rails.logger.warn "WOPI: searching by token #{token}"
+    User
+    .joins("LEFT OUTER JOIN tokens ON user_id = users.id")
+    .where('tokens.token = ?', token)
+    .where('tokens.ttl = 0 OR tokens.ttl > ?', Time.now.to_i)
+    .first
+  end
+
+  def get_wopi_token
+  	  # WOPI does not have a good way to request a new token, so a new token should be provided each time this is called, while keeping any old tokens
+  	  # as long as they have not yet expired
+
+  	tokens = Token.where("user_id = ?", id).distinct
+
+  	for token in tokens
+  		if (token.ttl < Time.now.to_i)
+  			token.delete
+  		end
+  	end
+
+    token_string = Devise.friendly_token(20) + "-"  + id.to_s
```

Comment:
Plus it might look better this way: `token_string = "#{Devise.friendly_token(20)}-#{id}"`.